Malte lig 535 scorer for keypoint detection

MalteEbner opened the pull request:
## Description
- Creates classes for representing keypoints of one object and all objects with keypoints in one image.
- Cretes a basic least confidence scorer.
- Includes unittests.

I did not add more advanced scorers yet, as this does not have high priority.

philippmwirth commented on lightly/active_learning/utils/keypoint_detection_output.py:
```diff
@@ -0,0 +1,60 @@
+            This is not used at the moment.
+
+    Examples:
+        >>> # Create the representation of wo keypoints with confidences
```
*two

philippmwirth commented on lightly/active_learning/utils/keypoint_detection_output.py:
```diff
@@ -0,0 +1,60 @@
+    Attributes:
+        keypoints:
+            Predicted keypoints in the [x0, y0, c0, ... xk, yk, ck] format.
+            Thus the shape of the array for k keypoints must be (k*3, ).
```
You could simply leave the "thus" away.

philippmwirth commented on lightly/active_learning/utils/keypoint_detection_output.py:
```diff
@@ -0,0 +1,60 @@
+        >>> keypoint_detections = KeypointDetection(keypoints, 3)
+    """
+
+    def __init__(self, keypoints: np.ndarray, object_id: int = -1):
```
is it required that the keypoints are an `np.ndarray`? They could also just be a list which is the more flexible typehint, no?

MalteEbner replied:
I changed it to a list.

philippmwirth commented on lightly/active_learning/utils/keypoint_detection_output.py:
```diff
@@ -0,0 +1,60 @@
+        """Raises a ValueError if the format is not as required.
+        """
+        if len(self.keypoints) % 3 != 0 or len(self.keypoints.shape) != 1:
+            raise ValueError("keypoints must be in the format of "
```
Use capital letter "K" and you can leave the "but they are not" away.

philippmwirth commented on lightly/active_learning/utils/keypoint_detection_output.py:
```diff
@@ -0,0 +1,60 @@
+        if len(self.keypoints) % 3 != 0 or len(self.keypoints.shape) != 1:
+            raise ValueError("keypoints must be in the format of "
+                             "[x0, y0, c0, ... xk, yk, ck], but they are not.")
+        if any(self.get_confidences() < 0):
```
Calling `get_confidences` repeatedly is inefficient

philippmwirth commented on lightly/active_learning/scorers/keypoint_detection.py:
```diff
@@ -0,0 +1,91 @@
+
+    Currently supports the following scorers:
+
+        `least_confidence`:
```
We should at least support `uncertainty_margin`

MalteEbner replied:
We don't have class distributions, thus `uncertainty_margin` and `uncertainty_entropy` don't make sense.

MalteEbner replied:
Thus also "least_confidence" is the wrong name, I renamed it to "mean_uncertainty" as it is `1 - mean(confidences over all keypoints)`.

philippmwirth commented on lightly/active_learning/scorers/keypoint_detection.py:
```diff
@@ -0,0 +1,91 @@
+        >>>         keypoint_detection = KeypointDetection(keypoints)
+        >>>         keypoint_detections.append(keypoint_detection)
+        >>>     output = KeypointDetectionOutput(keypoint_detections)
+        >>>     model_output.append(output)
```
the scorer part is missing 😄

philippmwirth commented on lightly/active_learning/scorers/keypoint_detection.py:
```diff
@@ -0,0 +1,91 @@
+    """
+    scores = []
+    for keypoint_detection_output in model_output:
+        confidences_this_detection = []
```
this name is very confusing, don't you mean something like `confidences_per_image` and then below the `confidences` are `confidences_per_instance`?

MalteEbner commented:
Additionally to addressing your PR comments, @philippmwirth , I have also renamed the objects to `KeypointPrediction` (per image) and `KeypointInstancePrediction` and added class methods to init from a dictionary or even directly a string. Now the prediction formats can be directly inserted:

```python
dicts = [
    {
        "category_id": 3,
        "keypoints": [423, 432, 0.4, 231, 655, 0.3],
        "score": -1.9
    }
]
KeypointPrediction.from_dicts(dicts)
```

https://my-organization-2.gitbook.io/top-secret-851/glossary/predictions#keypoint-detection

philippmwirth commented on lightly/active_learning/utils/keypoint_predictions.py:
```diff
@@ -0,0 +1,96 @@
+
+    """
+
+    def __init__(self, keypoint_instance_predictions: List[
```
In general I would propose we always use the same way to define functions:
If it all fits in one line:
```
def __init__(self, x, y):
```
otherwise
```
def __init__(
    self,
    x,
    y,
):
```

philippmwirth commented on lightly/active_learning/utils/keypoint_predictions.py:
```diff
@@ -0,0 +1,96 @@
+
+    @classmethod
+    def from_json_string(cls, json_string: str):
+        dicsts = json.loads(json_string)
```
typo: dicsts

philippmwirth replied:
also, the docstrings are missing

philippmwirth commented on lightly/active_learning/utils/keypoint_predictions.py:
```diff
@@ -0,0 +1,96 @@
+        self._format_check()
+
+    @classmethod
+    def from_dict(cls, dict_: Dict[str, Union[int, List[float], float]]):
```
This should most definitely have a docstring explaining the input format.

philippmwirth commented on lightly/active_learning/utils/keypoint_predictions.py:
```diff
@@ -0,0 +1,96 @@
+        self.keypoint_instance_predictions = keypoint_instance_predictions
+
+    @classmethod
+    def from_dicts(cls,
```
docstring is missing and would help a lot
